devito/ir/support/utils.py

```diff
@@ -0,0 +1,193 @@
+
+from devito.dimension import Dimension
+from devito.ir.support.basic import Access, Scope
+from devito.ir.support.space import Interval, Backward, Forward, Any
+from devito.ir.support.stencil import Stencil
+from devito.symbolics import retrieve_indexed
+from devito.tools import as_tuple, flatten
+
+__all__ = ['compute_intervals', 'detect_flow_directions', 'compute_directions',
+           'force_directions', 'group_expressions']
+
+
+def compute_intervals(expr):
+    """Return an iterable of :class:`Interval`s representing the data items
+    accessed by the :class:`sympy.Eq` ``expr``."""
+    # Deep retrieval of indexed objects in /expr/
+    indexeds = retrieve_indexed(expr, mode='all')
+    indexeds += flatten([retrieve_indexed(i) for i in e.indices] for e in indexeds)
+
+    # Detect the indexeds' offsets along each dimension
+    stencil = Stencil()
+    for e in indexeds:
+        for a in e.indices:
+            if isinstance(a, Dimension):
+                stencil[a].update([0])
+            d = None
+            off = [0]
+            for i in a.args:
+                if isinstance(i, Dimension):
+                    d = i
+                elif i.is_integer:
+                    off += [int(i)]
+            if d is not None:
+                stencil[d].update(off)
+
+    # Determine intervals and their iterators
+    iterators = OrderedDict()
+    for i in stencil.dimensions:
+        if i.is_NonlinearDerived:
+            iterators.setdefault(i.parent, []).append(stencil.entry(i))
+        else:
+            iterators.setdefault(i, [])
+    intervals = []
+    for k, v in iterators.items():
+        offs = set.union(set(stencil.get(k)), *[i.ofs for i in v])
+        intervals.append(Interval(k, min(offs), max(offs)))
+
+    return intervals, iterators
+
+
+def detect_flow_directions(exprs):
+    """Return a mapper from :class:`Dimension`s to iterables of
+    :class:`IterationDirection`s representing the theoretically necessary
+    directions to evaluate ``exprs`` so that the information "naturally
+    flows" from an iteration to another."""
+    exprs = as_tuple(exprs)
+
+    writes = [Access(i.lhs, 'W') for i in exprs]
+    reads = flatten(retrieve_indexed(i.rhs, mode='all') for i in exprs)
+    reads = [Access(i, 'R') for i in reads]
+
+    # Determine indexed-wise direction by looking at the vector distance
+    mapper = defaultdict(set)
+    for w in writes:
+        for r in reads:
+            if r.name != w.name:
+                continue
+            dimensions = [d for d in w.aindices if d is not None]
+            for d in dimensions:
+                try:
+                    if w.distance(r, d) > 0:
+                        mapper[d].add(Forward)
+                        break
+                    elif w.distance(r, d) < 0:
+                        mapper[d].add(Backward)
+                        break
+                    else:
+                        mapper[d].add(Any)
+                except TypeError:
```

Comment:
I probably understand this `if-then-else` wrong, but it looks like, for a given dimension,  it will choose the first direction it finds and stop (break), not checking the remaining satisfies it. Is it risky?

Reply:
Good question. So, here we're calculating the so called vector distance between **two** array accesses. In such a case, if you **have to** go forward/backward along a given axis ``i``, then you can go in **any** directions along any of the ``i + k, k > 0`` axes (ie, the inner dimensions are parallel). However, the actual direction of these ``i + k, k > 0`` axes may be dictated by other array accesses. So, short answer to your question: no, it's not risky.